src/NosCore.GameObject/Services/ExchangeService/ExchangeService.cs

```diff
@@ -0,0 +1,241 @@
+            {
+                session.SendPacket(session.Character.GenerateSay(
+                    Language.Instance.GetMessageFromKey(LanguageKey.EXCHANGE_BLOCKED, session.Account.Language),
+                    SayColorType.Purple));
+                return;
+            }
+
+            if (session.Character.IsRelatedToCharacter(targetSession.Character.VisualId, CharacterRelationType.Blocked))
+            {
+                session.SendPacket(new InfoPacket
+                {
+                    Message = Language.Instance.GetMessageFromKey(LanguageKey.BLACKLIST_BLOCKED,
+                        session.Account.Language)
+                });
+                return;
+            }
+
+            if (session.Character.InShop || targetSession.Character.InShop)
+            {
+                session.SendPacket(new MsgPacket
+                {
+                    Message =
+                        Language.Instance.GetMessageFromKey(LanguageKey.HAS_SHOP_OPENED, session.Account.Language),
+                    Type = MessageType.White
+                });
+                return;
+            }
+
+            session.SendPacket(new ModalPacket
+            {
+                Message = Language.Instance.GetMessageFromKey(LanguageKey.YOU_ASK_FOR_EXCHANGE,
+                    session.Account.Language),
+                Type = 0
+            });
+
+            targetSession.Character.SendPacket(new DlgPacket
+            {
+                YesPacket = new ExchangeRequestPacket
+                { RequestType = RequestExchangeType.List, VisualId = session.Character.VisualId },
+                NoPacket = new ExchangeRequestPacket
+                { RequestType = RequestExchangeType.Declined, VisualId = session.Character.VisualId },
+                Question = Language.Instance.GetMessageFromKey(LanguageKey.INCOMING_EXCHANGE, session.Account.Language)
+            });
+        }
+
+        public List<KeyValuePair<long, IvnPacket>> ProcessExchange(Tuple<long, long> users, IInventoryService sessionInventory, IInventoryService targetInventory)
+        {
+            var usersArray = new[] { users.Item1, users.Item2 };
+            var items = new List<KeyValuePair<long, IvnPacket>>(); //SessionId, PocketChange
+
+            foreach (var user in usersArray)
+            {
+                foreach (var item in _exchangeDatas[user].ExchangeItems)
+                {
+                    var destInventory = user == users.Item1 ? targetInventory : sessionInventory;
+                    var originInventory = user == users.Item1 ? sessionInventory : targetInventory;
+                    var targetId = user == users.Item1 ? users.Item2 : users.Item1;
+                    var sessionId = user == users.Item1 ? users.Item1 : users.Item2;
+                    var newItem = item.Value == item.Key.Amount ? originInventory.DeleteById(item.Key.Id) : originInventory.RemoveItemAmountFromInventory(item.Value, item.Key.Id);
```

Comment:
originInventory.DeleteById(item.Key.Id) you shouldnt delete and recreate. Look at https://github.com/NosCoreIO/NosCore/blob/master/src/NosCore.GameObject/Character.cs#L516